mountpoint-s3/src/checksums.rs

```diff
@@ -1,195 +1,4 @@
-use std::ops::RangeBounds;
-
-use bytes::{Bytes, BytesMut};
-use mountpoint_s3_crt::checksums::crc32c::{self, Crc32c};
-
-use thiserror::Error;
-
-/// A `ChecksummedBytes` is a bytes buffer that carries its checksum.
-/// The implementation guarantees that integrity will be validated before the data can be accessed.
-/// Data transformations will either fail returning an [IntegrityError], or propagate the checksum
-/// so that it can be validated on access.
-#[derive(Clone, Debug)]
-#[must_use]
-pub struct ChecksummedBytes {
-    orig_bytes: Bytes,
-    /// Always a subslice of `orig_bytes`
-    curr_slice: Bytes,
-    /// Checksum for `orig_bytes`
-    checksum: Crc32c,
-}
```

Comment:
Is it worth keeping this type around and using it exclusively as an internal detail of `ObjectPart`?

We would need to update this type with the changes you made (storing a range instead of a slice) and checksum over that (or maintaining the range twice maybe makes more sense), but maybe it keeps the `ObjectPart` simple as something that delegates to this for byte operations, and maintains a simple guard to the inner `ChecksummedBytes` by ensuring we provide the key, etc at access.

I'm imagining `ObjectPart` still taking `Bytes` as input and output, just keeping the checksumming of the bytes in this type.

Reply:
Just saw James' comment related w/ using `ChecksummedBytes` to protect the KMS key configuration when we add that. https://github.com/awslabs/mountpoint-s3/pull/644#issuecomment-1846337172

For a little context, I think we'd need to checksum an enum as well indicating which encryption mode to use (KMS SSE or KMS DSSE).

It feels like we have two patterns:
- Checksumming against one or more bits of data (byte ranges or otherwise... I guess they all end up as bytes in the end)
- Guarding said data by requiring the callee to provide metadata

Reply:
I suspect that moving the "byte operations" out into `ChecksummedBytes` may be complicated by `ObjectPart` also keeping track of the offset (in sync with the range) and maintaining the metadata checksum. 
On the other hand, the other usages of `ChecksummedBytes` we are considering, like the KMS example, likely will not require any operations on bytes and could use a much simpler data+checksum type.
But let me think a bit more about this.